libs/ui/src/hooks/use_smartcard_auth.ts

```diff
@@ -0,0 +1,272 @@
+            }
+            case 'error':
+              return { status: 'logged_out', reason: 'card_error' };
+            case 'ready': {
+              const loginResult = attemptLogin(
+                action.card,
+                allowedUserRoles,
+                electionDefinition
+              );
+              if (loginResult.isOk()) {
+                return { status: 'logged_in', user: loginResult.ok() };
+              }
+              return { status: 'logged_out', reason: loginResult.err() };
+            }
+            default:
+              throwIllegalValue(action.card, 'status');
+          }
+        })();
+        return { card: action.card, auth: newAuth };
+      }
+      default:
+        throwIllegalValue(action.type);
+    }
+  };
+}
+
+function buildCardStorage(
+  card: CardApiReady,
+  cardApi: Card,
+  cardLock: Lock
+): CardStorage {
+  return {
+    hasStoredData: card.longValueExists ?? false,
+
+    readStoredObject: async (schema) => {
+      await cardLock.waitForLock(LOCK_POLLING_INTERVAL);
+      const result = await cardApi.readLongObject(schema);
+      cardLock.unlock();
+      return result;
+    },
+
+    readStoredString: async () => {
+      await cardLock.waitForLock(LOCK_POLLING_INTERVAL);
+      try {
+        return ok(await cardApi.readLongString());
+      } catch (error) {
+        return err(error);
+      } finally {
+        cardLock.unlock();
+      }
+    },
+
+    readStoredUint8Array: async () => {
+      await cardLock.waitForLock(LOCK_POLLING_INTERVAL);
+      try {
+        return ok(await cardApi.readLongUint8Array());
+      } catch (error) {
+        return err(error);
+      } finally {
+        cardLock.unlock();
+      }
+    },
+
+    writeStoredData: async (value) => {
+      await cardLock.waitForLock(LOCK_POLLING_INTERVAL);
+      try {
+        if (value instanceof Uint8Array) {
+          await cardApi.writeLongUint8Array(value);
+        } else {
+          await cardApi.writeLongObject(value);
+        }
+        return ok();
+      } catch (error) {
+        return err(error);
+      } finally {
+        cardLock.unlock();
+      }
+    },
+
+    clearStoredData: async () => {
+      await cardLock.waitForLock(LOCK_POLLING_INTERVAL);
+      try {
+        await cardApi.writeLongUint8Array(Uint8Array.of());
+        return ok();
+      } catch (error) {
+        return err(error);
+      } finally {
+        cardLock.unlock();
+      }
+    },
+  };
+}
+
+export function useSmartcardAuth({
+  cardApi,
+  electionDefinition,
+  allowedUserRoles,
+}: UseSmartcardAuthArgs): SmartcardAuth {
+  const [{ card, auth }, dispatch] = useReducer(
+    smartcardAuthReducer(allowedUserRoles, electionDefinition),
+    {
+      card: { status: 'no_card' },
+      auth: { status: 'logged_out', reason: 'no_card' },
+    }
+  );
+  // Every card read/write must obtain a lock to prevent concurrent operations.
+  const cardLock = useLock();
```

Comment:
I'm interested in getting feedback on this idea. There was similar logic in useSmartcard, but I noticed that it only blocked concurrent writes and allowed reads during writes. Maybe that's fine? I'm not totally sure. Only allowing one card operation at a time seemed like a surefire way to ensure no concurrency issues.

Reply:
Thinking about it more, I'm wondering if concurrency protection is even necessary at all. I think consumers won't make concurrent long value reads/writes. The only concurrent requests I could imagine happening would be the short value reads from the useInterval. But the smartcard service will only process one request at a time, so I think those requests will get linearized by the server anyway.

Reply:
commenting just on this thing for now: having multiple layers of protection against concurrent requests is not a bad thing. I can imagine one day we change the backend service in some subtle way, and end up with concurrent requests processing... so I like the extra protection if it's not too onerous.

Reply:
Thanks, that's helpful. If it's just defensive and not solving an actual current problem, then I think it can be simplified. I changed it to better match the existing behavior of the locking in useSmartcard to just error on concurrent writes: 432c8725.